File: compiler/src/main/java/com/raizlabs/android/dbflow/processor/definition/ContentUriDefinition.java
package com.raizlabs.android.dbflow.processor.definition;

import com.raizlabs.android.dbflow.annotation.provider.ContentUri;
import com.raizlabs.android.dbflow.processor.Classes;
import com.raizlabs.android.dbflow.processor.model.ProcessorManager;

import javax.lang.model.element.Element;
import javax.lang.model.type.TypeMirror;

/**
 * Description:
 */
public class ContentUriDefinition extends BaseDefinition {

    public String name;

    public String classQualifiedName;

    public String path;

    public String type;

    public boolean queryEnabled;

    public boolean insertEnabled;

    public boolean deleteEnabled;

    public boolean updateEnabled;

    public ContentUri.PathSegment[] segments;

    public ContentUriDefinition(Element typeElement, ProcessorManager processorManager) {
        super(typeElement, processorManager);

        ContentUri contentUri = typeElement.getAnnotation(ContentUri.class);

        path = contentUri.path();

        type = contentUri.type();

        name = typeElement.getEnclosingElement().getSimpleName().toString()
                + "_" + typeElement.getSimpleName().toString();

        queryEnabled = contentUri.queryEnabled();

        insertEnabled = contentUri.insertEnabled();

        deleteEnabled = contentUri.deleteEnabled();

        updateEnabled = contentUri.updateEnabled();

        segments = contentUri.segments();

        TypeMirror typeMirror = typeElement.asType();
        if (!Classes.URI.equals(typeMirror.toString())) {
            processorManager.logError("Content Uri field returned wrong type. It must return a Uri");
        }
    }
}

File: processor/src/main/kotlin/com/dbflow5/processor/definition/provider/ContentUriDefinition.kt
package com.dbflow5.processor.definition.provider

import com.dbflow5.contentprovider.annotation.ContentUri
import com.dbflow5.processor.ClassNames
import com.dbflow5.processor.ProcessorManager
import com.dbflow5.processor.definition.BaseDefinition
import javax.lang.model.element.Element
import javax.lang.model.element.ExecutableElement
import javax.lang.model.element.VariableElement

/**
 * Description:
 */
class ContentUriDefinition(contentUri: ContentUri,
                           typeElement: Element, processorManager: ProcessorManager)
    : BaseDefinition(typeElement, processorManager) {

    var name = "${typeElement.enclosingElement.simpleName}_${typeElement.simpleName}"

    val path: String = contentUri.path
    val type: String = contentUri.type
    val queryEnabled: Boolean = contentUri.queryEnabled
    val insertEnabled: Boolean = contentUri.insertEnabled
    val deleteEnabled: Boolean = contentUri.deleteEnabled
    val updateEnabled: Boolean = contentUri.updateEnabled
    val segments = contentUri.segments

    init {
        if (typeElement is VariableElement) {
            if (ClassNames.URI != elementTypeName) {
                processorManager.logError("Content Uri field returned wrong type. It must return a Uri")
            }
        } else if (typeElement is ExecutableElement) {
            if (ClassNames.URI != elementTypeName) {
                processorManager.logError("ContentUri method returns wrong type. It must return Uri")
            }
        }
    }
}